### PR TITLE
Remove inclusion of `gmp.h` and `limits.h` in `flint.h`

### PR DESCRIPTION
**Pull request by albinahlback**
In order to clean up the name space and allow for faster compilations due to less complex header structures, remove these inclusions.

We now define FLINT_BITS during configuration, where it is pushed into config.h (or flint-config.h) and we set constants such as WORD_MAX according to this value.

Furthermore, define new types `nn_ptr` and nn_srcptr that is equivalent to `mp_ptr` and `mp_srcptr`.

FLINTify all types, but keep the `mpn_extras` and `fft` as is.

### CODE REVIEW

File: src/ulong_extras.h [hunk 1]
```diff
@@ -77,8 +77,14 @@ ulong n_flog(ulong n, ulong b);
...
 
-ulong __gmpn_gcd_11(ulong, ulong);
-ulong __gmpn_gcd_1(nn_srcptr, long int, ulong);
+#ifdef _MSC_VER
+# define DECLSPEC_IMPORT __declspec(dllimport)
+#else
+# define DECLSPEC_IMPORT
+#endif
+DECLSPEC_IMPORT ulong __gmpn_gcd_11(ulong, ulong);
+DECLSPEC_IMPORT ulong __gmpn_gcd_1(nn_srcptr, long int, ulong);
```
Comment (fredrik-johansson) [inline src/ulong_extras.h:86]: Should the second param really be ``long int`` here, not ``slong``?

Comment (albinahlback) [reply src/ulong_extras.h:86]: Yes, unless you are on a Cray supercomputer, `mp_size_t` is defined as `long int`.

**Comment by albinahlback** [comment]
Should be the final `mp_size_t` bug since that is the only function (of what I can see) in FLINT or GMP that takes a reference of an `mp_size_t` as argument.

**Comment by fredrik-johansson** [comment]
> Should be the final `mp_size_t` bug since that is the only function (of what I can see) in FLINT or GMP that takes a reference of an `mp_size_t` as argument.

Might be worth adding a comment.

**Pull request merged**